File: source/scene/Scene.java
package scene;

import editor.Editor;
import entity.Entity;
import entity.material.Material;
import glparts.Mesh;
import glparts.Texture;
import physics.abs.Physical;
import utility.Instance;
import utility.nncollection.NNArrayList;
import utility.nncollection.NNHashSet;

import java.io.*;
import java.util.Set;

public class Scene extends NNArrayList<Entity> implements Physical, Serializable {
    public final Set<Material> materials = new NNHashSet<>();
    public final Set<Texture> textures = new NNHashSet<>();
    public final Set<Mesh> meshes = new NNHashSet<>();

    public final SceneNameHolders names = new SceneNameHolders();
    public final SceneSelectedData selected = new SceneSelectedData();

    public Scene() {}

    @Override
    public boolean add(Entity entity) {
        if (super.add(entity)) {
            meshes.add(entity.components.mesh.mesh);
            materials.add(entity.components.material.material);

            if (Instance.isValid(entity.components.material.material)) {
                textures.add(entity.components.material.material.colorMap);
                textures.add(entity.components.material.material.normalMap);
                textures.add(entity.components.material.material.roughnessMap);
            }
            return true;
        }
        return false;
    }

    @Override
    public void updatePhysics(float deltaT) {
        for (Entity entity : this) {
            entity.updatePhysics(deltaT);
        }
    }

    public boolean toFile(String filepath) {
        try {
            boolean ignored = new File(filepath).getParentFile().mkdirs();
            try (ObjectOutputStream stream = new ObjectOutputStream(new FileOutputStream(filepath))) {
                stream.writeObject(this);
            }
            return true;
        }
        catch (Exception ignored) {
            System.out.println("Scene saving error, \"" + filepath + '"');
            return false;
        }
    }

    public static Scene fromFile(String filepath, Editor editor) {
        try (ObjectInputStream stream = new ObjectInputStream(new FileInputStream(filepath))) {
            Scene scene = (Scene) stream.readObject();
            for (Entity entity : scene) {
                entity.editor = editor;
            }
            return scene;
        }
        catch (Exception ignored) {
            System.out.println("Scene loading error, \"" + filepath + '"');
            return null;
        }
    }
}

File: source/scene/Scene.kt
package scene

import editor.Editor
import entity.Entity
import entity.material.Material
import glparts.Mesh
import glparts.Texture
import physics.abs.Physical
import java.io.*

class Scene : ArrayList<Entity>(), Physical, Serializable {
    val materials: MutableSet<Material> = HashSet()
    val textures: MutableSet<Texture> = HashSet()
    val meshes: MutableSet<Mesh> = HashSet()
    val names: SceneNameHolders = SceneNameHolders()
    val selected: SceneSelectedData = SceneSelectedData()

    override fun add(element: Entity): Boolean {
        if (super.add(element)) {
            element.components.mesh.mesh?.let {
                meshes.add(it)
            }
            element.components.material.material?.let {
                materials.add(it)
                it.colorMap?.let { it1 -> textures.add(it1) }
                it.normalMap?.let { it1 -> textures.add(it1) }
                it.roughnessMap?.let { it1 -> textures.add(it1) }
            }
            return true
        }
        return false
    }

    override fun updatePhysics(deltaT: Float) {
        for (entity in this) {
            entity.updatePhysics(deltaT)
        }
    }

    fun toFile(filepath: String): Boolean {
        return try {
            File(filepath).parentFile.mkdirs()
            ObjectOutputStream(FileOutputStream(filepath)).use { stream ->
                stream.writeObject(this)
            }
            true
        }
        catch (ignored: Exception) {
            println("Scene saving error, \"$filepath\"")
            false
        }
    }

    companion object {
        fun fromFile(filepath: String, editor: Editor): Scene? {
            try {
                ObjectInputStream(FileInputStream(filepath)).use { stream ->
                    val scene = stream.readObject() as Scene
                    for (entity in scene) {
                        entity.editor = editor
                    }
                    return scene
                }
            }
            catch (ignored: Exception) {
                println("Scene loading error, \"$filepath\"")
                return null
            }
        }
    }
}